utilities/xmlrpcpp/test/HelloTest.cpp

```diff
@@ -0,0 +1,258 @@
+
+#include <iostream>
+#include <stdlib.h>
+#include <boost/thread/thread.hpp>
+#include <boost/thread/mutex.hpp>
+#include <functional>
+
+#include <gtest/gtest.h>
+
+using XmlRpc::XmlRpcServerMethod;
+using XmlRpc::XmlRpcServer;
+using XmlRpc::XmlRpcClient;
+using XmlRpc::XmlRpcValue;
+
+// No arguments, result is "Hello".
+class Hello : public XmlRpcServerMethod {
+public:
+  Hello(XmlRpcServer* s) : XmlRpcServerMethod("Hello", s) {}
+
+  void execute(XmlRpcValue& params, XmlRpcValue& result) {
+    (void)params;
+    result = "Hello";
+  }
+
+  std::string help() {
+    return std::string("Say hello");
+  }
+};
+
+// One argument is passed, result is "Hello, " + arg.
+class HelloName : public XmlRpcServerMethod {
+public:
+  HelloName(XmlRpcServer* s) : XmlRpcServerMethod("HelloName", s) {}
+
+  void execute(XmlRpcValue& params, XmlRpcValue& result) {
+    std::string resultString = "Hello, ";
+    resultString += std::string(params[0]);
+    result = resultString;
+  }
+};
+
+// A variable number of arguments are passed, all doubles, result is their sum.
+class Sum : public XmlRpcServerMethod {
+public:
+  Sum(XmlRpcServer* s) : XmlRpcServerMethod("Sum", s) {}
+
+  void execute(XmlRpcValue& params, XmlRpcValue& result) {
+    int nArgs = params.size();
+    double sum = 0.0;
+    for (int i = 0; i < nArgs; ++i)
+      sum += double(params[i]);
+    result = sum;
+  }
+};
+
+class XmlRpcTest : public ::testing::Test {
+protected:
+  XmlRpcTest() : hello(&s), helloName(&s), sum(&s), port(0), done(false) {}
+
+  void work() {
+    while (!done) {
+      s.work(0.1); // run the worker queue for 100ms
+    }
+  }
+
+  virtual void SetUp() {
+    // XmlRpc::setVerbosity(5);
+
+    // Create the server socket. Passing 0 for the port number requests that
+    // the OS randomly select an available port.
+    s.bindAndListen(0);
+    // Retrieve the assigned port number.
+    port = s.get_port();
+
+    // Enable introspection.
+    s.enableIntrospection(true);
+
+    // Start the worker thread.
+    server_thread = boost::thread(boost::mem_fn(&XmlRpcTest::work), this);
+  }
+
+  virtual void TearDown() {
+    // TODO(austin): determine if we need to do anything here to avoid
+    // leaking resources
+    done = true;
+    server_thread.join();
+    s.shutdown();
+  }
+
+  // The server and its methods
+  XmlRpcServer s;
+  Hello hello;
+  HelloName helloName;
+  Sum sum;
+
+  // Server port number (for clients)
+  int port;
+
+  // Server thread
+  bool done;
+  boost::thread server_thread;
+};
+
+TEST_F(XmlRpcTest, Introspection) {
+  XmlRpcClient c("localhost", port);
+
+  // Use introspection API to look up the supported methods
+  XmlRpcValue noArgs, result;
+
+  ASSERT_TRUE(c.execute("system.listMethods", noArgs, result));
+
+  XmlRpcValue methods;
+  methods[0] = "Hello";
+  methods[1] = "HelloName";
+  methods[2] = "Sum";
+  methods[3] = "system.listMethods";
+  methods[4] = "system.methodHelp";
+  methods[5] = "system.multicall";
```

Comment:
I'm partial to `boost::assign::list_of` in these instances, but what's here is fine.